### ISSUE TITLE
Leading characters in wrapped lines are a little inconsistent

### ISSUE THREAD

**Issue by audreytoskin** [josh-berry/magic-reflow#2]
Magic Reflow applies a hanging indent to Markdown list items that are longer than a single line. For single-digit numbered list items, the hanging indent is 3 spaces, but for two-digit numbered list items, the hanging indent is a tab.

``` markdown
8. slkdfjskdl fskldfj lksj fklsj dfklsjd lfk sdklfj lksjd kfljsdklfj ljsdkfjlksd
   jflksj dfkl jsdf jlsjd fkljs dlkfjlsd jf
9. slkdfjskdl fskldfj lksj fklsj dfklsjd lfk sdklfj lksjd kfljsdklfj ljsdkfjlksd
   jflksj dfkl jsdf jlsjd fkljs dlkfjlsd jf
10. slkdfjskdl fskldfj lksj fklsj dfklsjd lfk sdklfj lksjd kfljsdklfj
	ljsdkfjlksd jflksj dfkl jsdf jlsjd fkljs dlkfjlsd jf
11. slkdfjskdl fskldfj lksj fklsj dfklsjd lfk sdklfj lksjd kfljsdklfj
	ljsdkfjlksd jflksj dfkl jsdf jlsjd fkljs dlkfjlsd jf
```

Perhaps because my editor is configured to normally indent with tabs and to render tabs as 4 characters wide? That is not consistent across all applications, or even all configurations of Atom, though. In my installation of Firefox, the tab characters in the above example appears to be 8 characters wide.

I've also noticed wrapping a Markdown blockquote produces different results depending on the format of the blockquote paragraph before it was wrapped: If I write a blockquote as a single line that is more than 80 characters long, Magic Reflow will wrap the paragraph with `> ` at the beginning of each line. But if I write a blockquote as a 2-line paragraph, with `> ` only at the beginning of the first line (which is still a valid Markdown blockquote), Magic Reflow does not include the right angle bracket at the beginning of lines when wrapping.

``` markdown
<!-- raw input -->
> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

<!-- wrapped output -->
> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
> tempor incididunt ut labore et dolore magna aliqua.

<!-- raw input -->
> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore
et dolore magna aliqua.

<!-- wrapped output -->
> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
tempor incididunt ut labore et dolore magna aliqua.
```

These inconsistencies seem like a bug to me, but I'm also opening Issue #3 because it's also possible users might want a choice as to which wrapping behavior to use.

**Comment by josh-berry** [comment]
We can keep this issue for your question about tabs and address the blockquote issue in #3.

The tabs-vs-spaces setting is controlled by the "Soft Tabs" setting in Settings > Editor. Similarly, Magic Reflow uses your "Tab Length" setting to determine how many spaces a tab character should count for.

So if "Soft Tabs" is turned off, and "Tab Length" is 4, Magic Reflow will insert one Tab character instead of 4 spaces.

There is, unfortunately, no universally right way to insert a tab, for the reason you mentioned above -- tabs appear as 4 spaces by default in Atom (and various other IDEs), but 8 spaces in many other places, and nobody can actually agree on how many spaces a tab character is.

So Magic Reflow--rightly or wrongly--follows Atom's conventions as configured in your editor.

If you want different behavior, ordinarily I would suggest toggling the "Soft Tabs" option for the language package where you're having the issue; however Atom doesn't seem to have a separate Markdown package (such that you could toggle it only for Markdown and not anywhere else). :/

One option might be to provide explicit "reflow with spaces" and "reflow with tabs" commands, which ignore the "Soft Tabs" setting entirely. Would this address your spaces-vs.-tabs issue?

**Comment by audreytoskin** [comment]
Yeah, it's tricky. I've been thinking about reworking the code style in my blog's Markdown pages and Pug templates, and the style / solution I'm currently leaning toward is using tabs for *block*-level indentation, and spaces for hanging indents and other cases where you need to align columns of text. Some would say mixing tabs and spaces is a mortal sin, but I think it works out fine when spaces' and tabs' purposes are clearly delineated like this. Spaces will always be more consistent, but tabs add fewer bytes to a file, and generally looks nicer when the editor has "show invisibles" enabled. I admit, my solution is probably unusual, but the debate always largely comes down to fickle taste, and this is what I've been thinking of lately.

So as far as spaces vs tabs, I actually like the current behavior: When I have hard tabs enabled, Magic Reflow preserves tab-indented blocks but space-indents the hanging lines of list items. It's just sufficiently long numbered lists, where it replaces the hanging spaces with a tab.

I might have expected to run into the opposite problem: I would have expected Magic Reflow to force *all* indentation to use either tabs or with spaces, and then I would have had to make the weird request of supporting a mix (or do the indents manually). Certainly there are plenty of people who would prefer to strictly enforce only one type of character for leading whitespace. Instead, for me the issue is that the reflow command doesn't differentiate between block indentation and hanging indentation.

**Comment by audreytoskin** [comment]
To sum up, my suggestion is this:

1. When wrapping lines, the block-level indentation (the whitespace before the list or sub-list's first `*` or `1.` item) should exactly duplicate the whitespace used on the first line *for the whole block*. Or else, for the rest of the lines of that list item. If the first item in a sublist starts with a tab or 7 spaces or whatever, all following lines within that sublist or list item should use the same leading whitespace. This would allow the user to do something different from normal if they feel they need to --- for example, maybe the user is writing a coding style guide and they want to talk about tabs vs spaces --- and yet this way it would also be easy to correct or alter the leading whitespace for a given block.
2. Any additional whitespace used for *hanging* indents on the list item's succeeding lines should always use either a variable number of spaces (so the text lines up), or a single tab.
3. Whether it uses spaces or tabs for hanging indents could by default be based on the editor's "soft tabs" setting, but should maybe be configurable.
